Enabling lite interpreter in torch python API

xcheng16 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39181 Enabling lite interpreter in torch python API**

Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

xcheng16 committed:
Enabling lite interpreter in torch python API

Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

[ghstack-poisoned]

xcheng16 committed:
Update on "Enabling lite interpreter in torch python API"


Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

[ghstack-poisoned]

dreiss commented on test/mobile/test_lite_script_module.py:
```diff
@@ -0,0 +1,38 @@
+
+import io
+
+from torch.jit.mobile import *
```
Please avoid `import *`, since that makes it hard to figure out where any given symbol comes from.

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+
+import os
+
+def load_for_lite_interpreter(f, map_location=None):
```
Can you put an underscore prefix on this so it doesn't become an externally-supported API?

xcheng16 replied:
sure.

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+def load_for_lite_interpreter(f, map_location=None):
+    r"""
+    Load a :class:`LiteScriptModule`
+    saved with :func:`torch.jit._save_for_lite_interpreter <torch.jit._save_for_lite_interpreter>`
```
Is it intentional that the name appears twice?

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+    Returns:
+        A :class:`LiteScriptModule` object.
+
+    Example:
+
+    .. testcode::
+
+        import torch
+        import io
+
+        # Load LiteScriptModule from saved file path
+        torch.jit._load_for_lite_interpreter('lite_script_module.pt')
+
+        # Load LiteScriptModule from io.BytesIO object
+        with open('lite_script_module.pt', 'rb') as f:
+            buffer = io.BytesIO(f.read())
+
+        # Load all tensors to the original device
+        torch.jit.mobile._load_for_lite_interpreter(buffer)
+
+    .. testcleanup::
+
+        import os
+        os.remove("lite_script_module.pt")
```
This seems like a little too much, but I'll leave it up to you.

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        import os
+        os.remove("lite_script_module.pt")
+    """
+    if isinstance(f, string_classes):
```
We don't support Python 2 anymore, so you can just use `str`.

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        import os
+        os.remove("lite_script_module.pt")
+    """
+    if isinstance(f, string_classes):
+        if not os.path.exists(f):
+            raise ValueError("The provided filename {} does not exist".format(f))
+        if os.path.isdir(f):
+            raise ValueError("The provided filename {} is a directory".format(f))
+
+    if isinstance(map_location, string_classes):
+        map_location = torch.device(map_location)
+    elif not (map_location is None or
+              isinstance(map_location, torch.device)):
+        raise ValueError("map_location should be either None, string or torch.device, "
+                         "but got type: " + str(type(map_location)))
+    if (str(map_location).startswith('cuda')):
+        validate_cuda_device(map_location)
```
If this code is duplicated from torch.jit, might want to pull it out into a method or two.

dreiss commented on torch/jit/mobile/__init__.py:
```diff
@@ -0,0 +1,86 @@
+        return self._c.forward(input)
+
+    def run_method(self, method_name, *input):
+        if method_name == '':
+            raise KeyError("method name can't be empty string \"\"")
```
Why check for this?  If this is an important constraint, doesn't the C++ code already validate it?

xcheng16 replied:
C++ will return null if it can not find a match, i think i can remove this check.

xcheng16 committed:
Update on "Enabling lite interpreter in torch python API"


Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

[ghstack-poisoned]

xcheng16 committed:
Update on "Enabling lite interpreter in torch python API"


Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

[ghstack-poisoned]

xcheng16 committed:
Update on "Enabling lite interpreter in torch python API"


Create a python binding classes torch._C. LiteScriptModule for mobile::module, a python class called LiteScriptModule is created which wrap torch._C. LiteScriptModule.
Python class LiteScriptModule contains preliminary functions including forward, run_method and __call__.

Create a python api "load_for_lite_interpreter" under torch.jit.mobile where takes pre-saved mobile module in a file-like object as input and returns python class LiteScriptModule.

Add a python binding method "_save_to_buffer_for_mobile" under ScriptModule, and python method "_save_to_buffer_for_lite_interpreter" under RecursiveScriptModule which saves mobile module into buffer instead of file.

Differential Revision: [D21757474](https://our.internmc.facebook.com/intern/diff/D21757474/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21757474/)!

[ghstack-poisoned]
